community/primitive-collections/src/main/java/org/neo4j/concurrent/AsyncEvents.java

```diff
@@ -0,0 +1,196 @@
+import java.util.concurrent.locks.LockSupport;
+
+import org.neo4j.function.Consumer;
+
+/**
+ * {@code AsyncEvents} is a mechanism for queueing up events to be processed asynchronously in a background thread.
+ *
+ * The {@code AsyncEvents} object implements {@link Runnable}, so it can be passed to a thread pool, or given to a
+ * dedicated thread. The runnable will then occupy a thread and dedicate it to background processing of events, until
+ * the {@code AsyncEvents} is {@link AsyncEvents#shutdown()}.
+ *
+ * If events are sent to an {@code AsyncEvents} that has been shut down, then those events will be processed in the
+ * foreground as a fall-back.
+ *
+ * Note, however, that no events are processed until the background thread is started.
+ *
+ * The {@code AsyncEvents} is given a {@link Consumer} of the specified event type upon construction, and will use it
+ * for doing the actual processing of events once they have been collected.
+ *
+ * @param <T> The type of events the {@code AsyncEvents} will process.
+ */
+public class AsyncEvents<T extends AsyncEvent> implements AsyncEventSender<T>, Runnable
+{
+    public interface Monitor
+    {
+        void eventCount( long count );
+
+        Monitor NONE = new Monitor()
+        {
+            @Override
+            public void eventCount( long count )
+            {
+            }
+        };
+    }
+
+    // TODO use VarHandles in Java 9
+    private static final AtomicReferenceFieldUpdater<AsyncEvents,AsyncEvent> STACK =
+            AtomicReferenceFieldUpdater.newUpdater( AsyncEvents.class, AsyncEvent.class, "stack" );
+    private static final AsyncEvent endSentinel = new Sentinel();
+    private static final AsyncEvent shutdownSentinel = new Sentinel();
+
+    private final Consumer<T> eventConsumer;
+    private final Monitor monitor;
+
+    @SuppressWarnings( {"unused", "FieldCanBeLocal"} )
+    private volatile AsyncEvent stack; // Accessed via AtomicReferenceFieldUpdater
+    private volatile Thread runner;
+    private volatile boolean shutdown;
+
+    /**
+     * Construct a new {@code AsyncEvents} instance, that will use the given consumer to process the events.
+     *
+     * @param eventConsumer The {@link Consumer} used for processing the events that are sent in.
+     */
+    public AsyncEvents( Consumer<T> eventConsumer, Monitor monitor )
+    {
+        this.eventConsumer = eventConsumer;
+        this.monitor = monitor;
+        this.stack = shutdownSentinel;
+    }
+
+    @Override
+    public void send( T event )
+    {
+        AsyncEvent prev = STACK.getAndSet( this, event );
+        assert prev != null;
+        event.next = prev;
+        if ( prev == endSentinel )
+        {
+            LockSupport.unpark( runner );
+        }
+        else if ( prev == shutdownSentinel )
+        {
+            AsyncEvent events = STACK.getAndSet( this, shutdownSentinel );
+            process( events );
+        }
+    }
+
+    @Override
+    public void run()
+    {
```

Comment:
Yeah. I think it is currently possible for the background thread to race with senders that think we've been shutdown. Senders with that impression will not call `LockSupport.unpark` on the background thread, which in turn can miss wake ups and go to sleep forever.

We can either fix this by spinning in the background thread to CAS `shutdownSentinel` to `endSentinel`, or we can initialise the `stack` field to `null` and add a guard clause to the beginning of the `send` method, that just process the event directly if the stack is observed to be `null`. Either approach would work, I think.